torch/testing/_internal/distributed/rpc/rpc_test.py

```diff
@@ -687,6 +687,24 @@ def test_profiler_with_sync_rpc_udf(self):
         self._profiler_test_with_rpc(RPCExecMode.SYNC, my_sleep_func, args=(1,),
                                      use_record_function=True)
 
+    @dist_init
+    def test_async_record_function(self):
+        num_sleep_seconds = 2
+        if self.rank == 1:
+            with torch.autograd.profiler.profile() as pf:
+                with torch.autograd.profiler.record_function("foo") as rf:
+                    fut = rpc.rpc_async(worker_name(0), my_sleep_func, args=(num_sleep_seconds,))
+                    rf._call_end_callbacks_on_future(fut)
+                # Note: calling fut.wait() outside of record_function to simulate
+                # real usage and to ensure that underlying handles in record_function
+                # are correctly persisted even after python ctx manager has exited
+                fut.wait()
+            # Find recorded event and roughly validate the time.
+            events = pf.function_events
+            rpc_event = [event for event in events if "foo" in event.name][0]
+            self.assertGreaterEqual(rpc_event.cpu_time_total * 1e-6, num_sleep_seconds)
```

Comment:
Would be also nice to add a test where we don't call `_call_end_callbacks_on_future` and validate that the reported time is not accurate and only includes the "async" dispatch part. The check can be cpu_time_total < num_sleep_seconds (since its hard to estimate the actual async dispatch time).

Reply:
@pritamdamania87 I don't think that with the current implementation that will work unfortunately. We still have the previous profiling with the `RecordFunction` that I implemented a while ago, so the following will happen:

1. Push "foo" record_function onto stack
2. Push RPC profiling key onto stack (from the previous profiling implementation)
3. Python record_function from 1 exits, results in a pop, but this pop gets incorrectly associated with (2)
4. Previous profiling implementation results in a pop when executing the future's callbacks, this gets associated with (1) incorrectly - they are inverted

To fix this, I'll update the diff to remove the old profiling approach since we will be replacing it anyway with this new one.